### ISSUE TITLE
Feature request : Import / Export file formats (CSV, JSON)

### ISSUE THREAD

**Issue by estellecomment** [nocodb/nocodb#179]
Hello ! 

My data is in a huge csv file (2M lines). I would like to import it into a table.
Ideally it would auto-create a table, taking the column names from the first line of CSV.
I would like to upload the file from my local disk, from my web browser.

As a bonus, it would be nice to have flexibility on the CSV format : my file is using "|" as a separator rather than "," 
(but I could transform my data before uploading it, so it's not a hard requirement. But it's preferable because I would like minimize the steps that I have to do in my workflow.)

Thanks for this cool product !

**Comment by o1lab** [comment]
CSV import/export is a good feature. We would support it.

However 2M lines/records is kind of ETL space Im afraid. Doing it over a network I've seen even mysql workbench struggles with AWS RDS. Talend and such are nice tools for this.

What tools have you considered before ?

**Comment by lifehome** [comment]
Hi,

First of all, this product looks amazing, and I really look forward to seeing it grow as time flies!

Cut to the point, I am the Community Manager maintaining the daily ops of [wars.vote4.hk](https://wars.vote4.hk/en)([Github repo here](https://github.com/vote4hk/warsinhk)) for the past year and still going on. We primarily use Google Spreadsheet to store all kinds of data, and export the sheets in CSV form, then transform and mutate it for GatsbyJS to display.

So the data existences can be in forms of:
- CSV (Endpoint provided by public Google Spreadsheet API)
- MS Excel (and all sorts of extensions Google Spreadsheet can export)
- JSON (The way we mutated the data, it can be key-value pair[(example here)](https://covidapi.vote4.hk) or formatted for GraphQL insertion)

For now, using Google Spreadsheet is great, it's free and comes with great collaboration, but the application(Google Spreadsheet) itself simply not designed to hold very large dataset. We have more than 20k rows for patient location tracking data point, and 10k more rows about confirmed cases, and it lags whenever a volunteer opens up the spreadsheet **before** editing.

If this feature becomes available, either CSV or JSON with formatter(to parse which line would be the indexing row), this would drastically allow multiple dataset to be imported via either GUI or API. If there is such thing as join tables, we can immediately pull up another Grid View and cross-reference multiple rows.

Again, really great work on this product, lovely, stunning and very powerful. Just spun up a test docker to try out and immediately fall in love with it, can't wait to see what it can be done in the future!

Thanks! Stay safe and healthy!

Regards,
Ivan

**Comment by estellecomment** [comment]
> CSV import/export is a good feature. We would support it.

Cool ! Thanks !

> However 2M lines/records is kind of ETL space Im afraid. Doing it over a network I've seen even mysql workbench struggles with AWS RDS. Talend and such are nice tools for this

Yes, I started thingking about this as well... For this data, the most important thing is to get an API to search in the data. So I am going towards https://github.com/jrieke/fastapi-csv, a package built on top of fastapi, a python API maker. It's more dev work but for the large data size it does look better.

I work in a startup incubator and was hoping that a shared instance of nocodb could be helpful for prototyping in neighboring projects. So even though I am not using it for this data, we will probably still run an instance, and see what others do with it :)

**Comment by lifehome** [comment]
Just to add, I think the ETL bottleneck happens all the time when data has become a gigantic set.
Rather of rendering all the data at once or in batch, I suggest default to pagination and a toggle in between showing all.
(Maybe just open up the page last time they left)

Anyway I think these problems or relevant performance issues can be contained in another ticket. 😄 

-- Ivan

**Comment by tegansnyder** [comment]
My team is just looking to export data to CSV or XLSX. We can handle imports directly via ETL jobs to the database.

**Comment by flavienbwk** [comment]
This should be set as high priority, I can't imagine using NocoDB without this feature :)

**Comment by armond1978** [comment]
Is there any ETA for this feature?

**Comment by o1lab** [comment]
Export grid view as CSV is available [with version 0.81.1](https://github.com/nocodb/nocodb/releases/tag/0.81.1)

Upload we will look into later.

**Comment by flavienbwk** [comment]
Thanks for the news.

NocoDB being a capitalization tool, upload would be _useful_ for migrating to it.

**Comment by o1lab** [comment]
Will use #730 for importing data.

**Comment by fishnux** [comment]
Can CSV export be used through the REST API? Anyone knows how?

**Comment by apprentiLAB-Suzanne** [comment]
Hi, I was testing CSV export / reimport, and I am having some trouble with field types. I tried exporting and reimporting one of my tables with a few different field types (URL, CheckBox, MultipleSelect, LongText...) and I noticed the field types did not match my original table. I saw I could assign field types myself when importing, but it could be a bit of a hassle for someone new to the tool. Looking at the CSV file I understand that field types are not exported in it.

So I was wondering if JSON export was a considered feature ? And if it could provide a more complete way to transfer / restore / transfer data ?

Anyway, thanks for all the great work :)
